dirwatcher.py

```diff
@@ -0,0 +1,204 @@
+
+# logging config
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s: %(levelname)s : %(message)s')
+file_handler = logging.FileHandler('dirwatcher.log', 'a')
+file_handler.setFormatter(formatter)
+logger.addHandler(file_handler)
+
+# add stream to include Traceback in logs
+stream_handler = logging.StreamHandler()
+stream_handler.setFormatter(formatter)
+logger.addHandler(file_handler)
+
+# exit option
+close_program = False
+
+
+def action_signal(sig_num, frame):
+    '''Actions incoming signals and closes program if needed'''
+    logger.warning(f'INCOMING SIGNAL - {signal.Signals(sig_num).name}')
+    global close_program
+    close_program = True
+
+
+# handle interval
+interval = 0
+
+# program globals
+current_dict = {}
+state = {}
+
+
+def watch_dir(dir_, search_text, ext=None):
+    '''Loops thru filenames in a dir & searches for magic str.
+        If filename already in dic, updates lineno, if not in dict, gives
+        default value of 0. Returns dict
+    '''
+    # global current_dict
+    if os.path.isdir(dir_):
+        dir_files = []
+
+        for dirpath, _, filenames in os.walk(dir_):
+            for f in filenames:
+                dir_files.append(os.path.abspath(os.path.join(dirpath, f)))
+        if not ext:
+            # if ext is NOT set
+            for file_ in dir_files:
+                # if not in current_dict, make key and set to 0
+                if file_ not in current_dict:
+                    current_dict[file_] = 0
+
+        elif ext:
+            # if ext IS set
+            for file_ in dir_files:
+                file_ending = file_[file_.rindex('.'):]
+                # and file has correct extension
+                if file_ending == ext:
+                    # and not in current_dict, add to current_dir
+                    if file_ not in current_dict:
+                        current_dict[file_] = 0
+
+        # parse current_dict for search_text and update
+        search_dict(current_dict, search_text)
+
+        # compare state_dict w/ current_dict and log changes
+        global state
+        compare(state=state, dict=current_dict)
+        # update state
+        state = current_dict
+
+
+def search_dict(dict_, text):
+    '''Searches dict files for text.  Stores last line searched and
+        Returns updated dict. Updates state.
+    '''
+    if dict_:
+        line_count = 0
+        for path, value in dict_.items():
+            line_count = value
+            with open(path) as f:
+                for line in f.readlines()[line_count:]:
+                    line_count += 1
+                    if text in line:
+                        logger.info(
+                            f'Found "{text}" in "{path}" (Line {line_count})')
+            dict_[path] = line_count
+        return dict_
+
+    else:
+        logger.info('Directory has no searchable files!')
+
+
+def compare(state=state, dict=current_dict):
```

Comment:
What happens if I simply change the name of a file being watched?  this will neither grow nor shrink the dictionary ...